ra/ra.go

```diff
@@ -1717,26 +1727,28 @@ func (ra *RegistrationAuthorityImpl) recordValidation(ctx context.Context, authI
 	if err != nil {
 		return err
 	}
-	var expires int64
+	var expires time.Time
 	if challenge.Status == core.StatusInvalid {
-		expires = authExpires.UnixNano()
+		expires = *authExpires
 	} else {
-		expires = ra.clk.Now().Add(ra.authorizationLifetime).UnixNano()
+		expires = ra.clk.Now().Add(ra.authorizationLifetime)
 	}
 	vr, err := bgrpc.ValidationResultToPB(challenge.ValidationRecord, challenge.Error)
 	if err != nil {
 		return err
 	}
-	var validated int64
+	var validated time.Time
 	if challenge.Validated != nil {
-		validated = challenge.Validated.UTC().UnixNano()
+		validated = challenge.Validated.UTC()
 	}
 	_, err = ra.SA.FinalizeAuthorization2(ctx, &sapb.FinalizeAuthorizationRequest{
 		Id:                authzID,
 		Status:            string(challenge.Status),
-		ExpiresNS:         expires,
+		ExpiresNS:         expires.UnixNano(),
+		Expires:           timestamppb.New(expires),
 		Attempted:         string(challenge.Type),
-		AttemptedAtNS:     validated,
+		AttemptedAtNS:     validated.UnixNano(),
+		AttemptedAt:       timestamppb.New(validated),
```

Comment:
I don't think this one works. Previously, we were setting `AttemptedAtNS` to be the zero-value of an int64, i.e. 0. Now we're setting it to be the UnixNano() of the zero-value of a time.Time, and that's not the same thing:
https://go.dev/play/p/aC9XiSTbHnZ
```
> fmt.Println(time.Time{}.UnixNano())
-6795364578871345152
```
So I think this one needs to have logic more like you were using up in pb-marshalling.go, where you use two separate variables for the old and new styles, and ensure that both have their zero value (i.e. the integer zero, and a nil pointer) if we want them to be unset.

Reply:
Good catch.